[MRG] Implement 'Translation Matrix'

robotcator opened the pull request:
This PR adds translation matrix feature to gensim model.

gojomo commented:
Very excited to see this! Note that it may be useful for more than just sets of word-vectors – for example, if you were to Doc2Vec-train a batch of new documents, but include a bunch (100s? 1000s? TBD) of reference documents from a prior larger run, you could presumably translate the new doc-vecs back into the older/larger space using those reference docs to learn the back-mapping. 

The comments & documentation need a lot of review/improvement for clarity/completeness. (Variable and method names might benefit from some careful consideration, too.) When it's thought to be ready for merging – passing tests and feature complete – I'll be happy to give it a close top-to-bottom read.

robotcator commented:
@gojomo Thanks, happy to have you help. You mean the translation matrix not only restrict to translate the word, but can apply to the document translation? Am I right?

gojomo commented:
Not document 'translation' as in from one natural-language to another, but from alternate Doc2Vec runs on different sets of (same-language) documents - but featuring enough overlapping texts to learn the mapping.

robotcator commented:
You mean when I were to train on a batch of new documents, and get the new doc vector. And There are also many reference documents, we can use those to learn to back-mapping. But I don't quite understand the reference documents( in same language), can you explain in more detail? Thanks.

gojomo commented:
Let's say you train doc-vecs for 10 million documents. Then you get a batch of 500,000 more. 

If you train the old model on just the new 500K, the model is improving for those – but drifting away from modelling the older docs. (If some docs/doc-types are not re-presented, in interleaved fashion, the model will lose its 'tuning' for them.)

Alternatively, if you train a new model on just the new 500K, the doc-vectors from this new model will have no essential relationship with the older model. (There's no absolute vector positioning from any training session - the exact same document can wind up in very different places based on the inherent randomness of initialization and training. It's only the *relative* positions of doc-vectors from the same combined training that is meaningful.)

So, I'm proposing it could be effective to train the new 500K, with some small set of "reference point" documents from the older set mixed-in. (Let's say for now: 10,000 docs.) When the training of the all-new model is done, those 10K ref docs will have points in the "new" space that render them comparable to the 500K new docs. These points will quite probably be very *unlike* their points in the older space. But, using this `TranslationMatrix`, those 10K points can be used to learn a mapping back to the older (10M doc) space, and then transform all 500K new docs into the original, older space for comparison to any of the now 10.5M documents.

robotcator commented:
Thank for you explaining. I understand what you said now. The "reference documents" from the older set will be mixed into the new training set, those can be seen as pivot to learn the "TranslationMatrix". It is worth trying.

robotcator commented:
Hi, @gojomo recently, I have doing the experiment about produce new document vectors via the translation matrix which learned by the back-mapping method. If I train on new 500K document , with 10k of "reference point" documents taken from the 10 million docs. Then, by using TranslationMatrix, we can transform all 500K new docs into the original, older space. So we got 500k more document vector in the original space.  But Is there any need measure/evaluate the new docs' vector in older space (perform better)? Or we just use this method to produce more doc vectors rather than train the entire corpus(10.5 million document )again?

piskvorky commented:
Hi guys, what's the context here? Can you please add some description of the purpose of this PR? 

What are its use-cases, from gensim users' POV?

robotcator commented:
@piskvorky This PR is to implement the [translation matrix](https://github.com/RaRe-Technologies/gensim/wiki/GSOC-2017-project-ideas). If needed, I will add more description in this PR.

gojomo commented:
@robotcator - Yes, an experiment to evaluate the usefulness of such reference-projected doc-vectors would be good – especially in comparison to inferring-doc-vectors. (Maybe inference works better. Or maybe if there are a lot of new words in the incremental docs, the back-projection works better... it'd certainly be able to bootstrap word-vectors, in the original space, that inference alone would not.) But, I don't have a proposed experiment design ready.

robotcator commented:
Hi, I use the  IMDB sentiment dataset to train two models, one with 15k documents and the other with 25k documents, all other parameters are the same. The code is below(some are from the notebook)
```
import gensim
from gensim.models.doc2vec import TaggedDocument
from gensim.models import Doc2Vec
from collections import namedtuple
from gensim import utils

def read_sentimentDocs():
    SentimentDocument = namedtuple('SentimentDocument', 'words tags split sentiment')

    alldocs = []  # will hold all docs in original order
    with utils.smart_open('alldata-id.txt', encoding='utf-8') as alldata:
        for line_no, line in enumerate(alldata):
            tokens = gensim.utils.to_unicode(line).split()
            words = tokens[1:]
            tags = [line_no] # `tags = [tokens[0]]` would also work at extra memory cost
            split = ['train','test','extra','extra'][line_no//25000]  # 25k train, 25k test, 25k extra
            sentiment = [1.0, 0.0, 1.0, 0.0, None, None, None, None][line_no//12500] # [12.5K pos, 12.5K neg]*2 then unknown
            alldocs.append(SentimentDocument(words, tags, split, sentiment))

    train_docs = [doc for doc in alldocs if doc.split == 'train']
    test_docs = [doc for doc in alldocs if doc.split == 'test']
    doc_list = alldocs[:]  # for reshuffling per pass

    print('%d docs: %d train-sentiment, %d test-sentiment' % (len(doc_list), len(train_docs), len(test_docs)))

    return train_docs, test_docs, doc_list

train_docs, test_docs, doc_list = read_sentimentDocs()

import multiprocessing
from random import shuffle

cores = multiprocessing.cpu_count()
# PV-DM w/concatenation - window=5 (both sides) approximates paper's 10-word total window size
model = Doc2Vec(dm=1, dm_concat=1, size=100, window=5, negative=5, hs=0, min_count=2, workers=cores)

small_train_docs = train_docs[:15000]

# train for small corpus
model.build_vocab(small_train_docs)
for epoch in xrange(50):
    shuffle(small_train_docs)
    model.train(small_train_docs, total_examples=len(small_train_docs), epochs=1)
# save the model
model.save("small_doc_15000_iter50.bin")


# train for large corpus
model.build_vocab(train_docs)
for epoch in xrange(50):
    shuffle(train_docs)
    model.train(train_docs, total_examples=len(train_docs), epochs=1)
# save the model
model.save("large_doc_25000_iter50.bin")
```
I use those 15k documents as reference document to mix in the new 15k documents, so I got the model2 with 25k documents. Then using those 15k documents and their corresponding doc vectors in model1 and model2, we can train a translation matrix. The code is below:
```
# coding=utf-8
from gensim.models import Doc2Vec
from sklearn.linear_model import LogisticRegression
from collections import namedtuple

import numpy as np

model1 = Doc2Vec.load("small_doc_15000_iter50.bin")
model2 = Doc2Vec.load("large_doc_25000_iter50.bin")

l = model1.docvecs.count
l2 = model2.docvecs.count
m1 = np.array([model1.docvecs[i] for i in range(l)])
m2 = np.array([model2.docvecs[i] for i in range(l)])

# train the translation matrix
tm = np.linalg.lstsq(m2, m1, -1)[0]
# back-mapping for the 15k documents
model1_15k_docvec = []
for i in range(l, l2):
    x = model2.docvecs[i]
    y = np.dot(x, tm)
    model1_15k_docvec.append(y)
```
The problem is how to evaluate the 15k document vector learned by the method, I think we can train a classifier to see whether those vector works. But we need a benchmark for comparison. The code is rough but did I explain the problem clear?@menshikh-iv

piskvorky commented on gensim/models/translation_matrix.py:
```diff
@@ -208,8 +208,8 @@ def translate(self, source_words=None, topn=5, additional=None, source_lang_vec=
 
-        if not isinstance(source_words, [string_types, list]):
-            raise ValueError("The word to be translated must be string type or a list of string.")
+        if not isinstance(source_words, (string_types, list, tuple)):
```
This looks rather unpythonic; why is such test needed?

Python uses duck-typing.

robotcator replied:
I see, will remove this type checking.

menshikh-iv commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,258 @@
+        self.source_lang_vec = source_lang_vec
+        self.target_lang_vec = target_lang_vec
+
+        self.translation_matrix = None
```
useless line

menshikh-iv commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,258 @@
+
+        self.translation_matrix = None
+
+        self.source_space = self.build_space(self.source_lang_vec, set(self.source_word))
```
Please call `Space.build(lang_vec, words)` directly

menshikh-iv commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,258 @@
+
+        self.translation_matrix = self.train(self.source_space, self.target_space)
+
+    def build_space(self, lang_vec, words=None):
```
useless method (according to below comment)

menshikh-iv commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,258 @@
+        hubness problem." arXiv preprint arXiv:1412.6568 (2014).
+        """
+
+        if source_words is None:
```
If this argument is mandatory, please don't use `kwargs`

menshikh-iv commented on gensim/test/test_translation_matrix.py:
```diff
@@ -0,0 +1,86 @@
+        target_word_vec_file = datapath("IT.1-10.cbow1_wind5_hs0_neg10_size300_smpl1e-05.txt")
+        target_word_vec = KeyedVectors.load_word2vec_format(target_word_vec_file, binary=False)
+
+        translation_matrix.TranslationMatrix(word_pair, source_word_vec, target_word_vec)
```
You create a translation matrix but don't use it, please add the translation as a test. Also add basic check (for translation matrix size for example)

menshikh-iv commented on gensim/test/test_translation_matrix.py:
```diff
@@ -0,0 +1,86 @@
+
+        test_word_pair = [("one", "uno"), ("two", "due")]
+        test_source_word, test_target_word = zip(*test_word_pair)
+        transmat.translate(test_source_word, topn=3)
```
Add an assert (check that your translation is correct)

menshikh-iv commented on gensim/test/test_translation_matrix.py:
```diff
@@ -0,0 +1,86 @@
+        test_source_word, test_target_word = zip(*test_word_pair)
+        transmat.translate(test_source_word, topn=3)
+
+    def test_translate_GC(self):
```
What's a main difference between NN and GC?

robotcator replied:
Given a source element x, the NN method to retrieve a target space label y is to returning the most similar word (according to the cosine similarity in this). The GC method is to return the target space label y which has x  ranked highest. The implementation of GC can computed as :  
GC(x, T) = arg min (Rank_{y, P} （x） - cos(x, y) ), where x is the source element to be translated, T the target retrieval space, P is set of elements sampled from the source space. Sorry for it didn't support Latex format. You can refer this paper in formulation 5. [website](https://arxiv.org/pdf/1412.6568.pdf)

menshikh-iv commented on gensim/test/test_translation_matrix.py:
```diff
@@ -0,0 +1,86 @@
+
+        test_word_pair = [("one", "uno"), ("two", "due")]
+        test_source_word, test_target_word = zip(*test_word_pair)
+        transmat.translate(test_source_word, topn=3, additional=10)
```
Add an assert (check that your translation is correct)

menshikh-iv commented on docs/notebooks/translation_matrix.ipynb:
Please add non-trivial examples (fruits, animals, random words, etc). Also please add your experiments in the distinct notebook.

robotcator commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,258 @@
+        """
+        self.mat = matrix
+        self.index2word = index2word
+        self.build_word2index()
```
Yes, I found that the Exception check is redundant and will remove it.

robotcator commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,258 @@
+            space object for the lexicon
+        """
+        if lang_vec is None:
+            raise RuntimeError("the word vector must be provided!")
```
I see, I will fix this here and below.

menshikh-iv commented on docs/notebooks/Translation Matrix Revisit.ipynb:
```diff
@@ -0,0 +1,341 @@
+    "            tokens = gensim.utils.to_unicode(line).split()\n",
+    "            words = tokens[1:]\n",
+    "            tags = [line_no] # `tags = [tokens[0]]` would also work at extra memory cost\n",
+    "            split = ['train','test','extra','extra'][line_no//25000]  # 25k train, 25k test, 25k extra\n",
```
Missing spaces

menshikh-iv commented on gensim/test/test_translation_matrix.py:
```diff
@@ -0,0 +1,93 @@
+
+        transmat = translation_matrix.TranslationMatrix(word_pair, source_word_vec, target_word_vec)
+
+        test_word_pair = [("one", "uno"), ("two", "due")]
```
Add more word pairs (and maybe not only numbers)

menshikh-iv commented on gensim/test/test_translation_matrix.py:
```diff
@@ -0,0 +1,93 @@
+
+        self.assertTrue(np.allclose(transmat.translation_matrix, loaded_transmat.translation_matrix))
+
+    def test_translate_NN(self):
```
lower case for method names (here and below)

menshikh-iv commented on gensim/test/test_translation_matrix.py:
```diff
@@ -0,0 +1,93 @@
+
+class TestTranslationMatrix(unittest.TestCase):
+    def test_translation_matrix(self):
+        train_file = datapath("OPUS_en_it_europarl_train_one2ten.txt")
```
Please load all needed models/files only once in `setUp` method (because you re-use this several times)

menshikh-iv commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,257 @@
+from gensim import utils
+from six import string_types
+
+random.seed(2333)
```
Please don't hardcode global random seed. You should create personal `Random` object and use it

menshikh-iv commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,257 @@
+        self.source_space = Space.build(self.source_lang_vec, set(self.source_word))
+        self.target_space = Space.build(self.target_lang_vec, set(self.target_word))
+
+        self.translation_matrix = self.train(self.source_space, self.target_space)
```
Don't run `train` in `__init__` method.

menshikh-iv commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,257 @@
+        # if the language word vector not provided by user, use the model's
+        # language word vector as default
+        if source_lang_vec is None:
+            warnings.warn("the parameter source_lang_vec didn't specified,"
```
We ignore line size limitation, please write all comment in one line (here and below). Also, use full stops and capitals where your start your sentence.

menshikh-iv commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,257 @@
+        """
+        return Space(np.dot(words_space.mat, self.translation_matrix), words_space.index2word)
+
+    def translate(self, source_words, topn=5, additional=None, source_lang_vec=None, target_lang_vec=None):
```
please add argument for algorithm (`additional=None` looks strange)

piskvorky commented on gensim/models/translation_matrix.py:
```diff
@@ -181,16 +182,17 @@ def apply_transmat(self, words_space):
+            gc: defines the training algorithm. By default (`gc=0`), use standard NN retrieval.
+            Otherwise use globally corrected neighbour retrieval method(as described in[1]).
+            additional: additional is an int that specify the number of word to sample from the source lexicon.
+            if gc=1, then additional must be provided.
```
Additional what?

robotcator replied:
Sorry for missing understanding and I will use another word. The parameter 'additional' is the number of words which sampled from the source language(usually the vocabulary in the source language's word2vec object) for the globally corrected neighbor retrieval method.

piskvorky replied:
Ah, I see. You meant literal `additional`.

It's best to format code as literals (using ``` backticks), to avoid misunderstandings like this.

piskvorky commented on gensim/models/translation_matrix.py:
```diff
@@ -209,20 +211,20 @@ def translate(self, source_words, topn=5, additional=None, source_lang_vec=None,
-        if additional is not None:
+        if gc:
+            if additional is None:
+                raise RuntimeError("When using the globally corrected neighbour retrieval method, the additional parameter which the number of words sampled from source space must be provided.")
```
There seems to be a word missing after `which` -- the sentence doesn't make sense now.

robotcator replied:
Thanks, I will revise it. And if there is any other inappropriate expression, please point out.

menshikh-iv commented on docs/notebooks/Translation Matrix Revisit.ipynb:
```diff
@@ -0,0 +1,344 @@
+{
```
Remove spaces in the filename `Translation_Matrix_Revisit.ipynb`

menshikh-iv commented on docs/notebooks/Translation Matrix Revisit.ipynb:
```diff
@@ -0,0 +1,344 @@
+    "    train_array[i] = m2[i]\n",
+    "    train_label[i] = 1\n",
+    "\n",
+    "    train_array[i+12500] = m2[i+12500]\n",
```
Spaces before/after `+` (here and anywhere)

menshikh-iv commented on docs/notebooks/Translation Matrix Revisit.ipynb:
```diff
@@ -0,0 +1,344 @@
+     ]
+    }
+   ],
+   "source": [
```
Need more explanation between cells (what happens)

menshikh-iv commented on docs/notebooks/translation_matrix.ipynb:
Please use plotly for all viz

robotcator replied:
done

menshikh-iv commented on docs/notebooks/translation_matrix.ipynb:
```diff
@@ -0,0 +1,1352 @@
+{
```
Also please add a description before each image (what I see now).

menshikh-iv commented on docs/notebooks/translation_matrix.ipynb:
E251 unexpected spaces around keyword / parameter equals (here and anywhere)

menshikh-iv commented on docs/notebooks/translation_matrix.ipynb:
E226 missing whitespace around arithmetic operator (here and anywhere)

robotcator replied:
Done

menshikh-iv commented on gensim/test/test_translation_matrix.py:
```diff
@@ -0,0 +1,118 @@
+    def test_translation_matrix(self):
+        model = translation_matrix.BackMappingTranslationMatrix(self.train_docs[:5], self.source_doc_vec, self.target_doc_vec)
+        transmat = model.train(self.train_docs[:5])
+        self.assertEqual(transmat.shape, (100, 100))
```
Why only "sanity check"? Please add another tests (similar with translation tests from previous class)?

robotcator replied:
This is a little different with the translation of word. We build the mapping `F` between the target model and original model(old). Given a document vector `x` from the target model,  we infer the vector by multiple `F` with `x` to get a new vector for the original vector. By now, we didn't have a good evaluation metric to assess the vector. So I only give the "sanity check" now.

menshikh-iv replied:
You can calculate cosine distance between vector from model and vector from translation

menshikh-iv commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,329 @@
+        return translated_word
+
+
+class BackMappingTranslationMatrix(utils.SaveLoad):
```
Please add example with viz for it (to translation matrix notebook)

menshikh-iv commented on docs/notebooks/Translation_Matrix_Revisit.ipynb:
```diff
@@ -0,0 +1,366 @@
+{
```
Please add this test to `translation_matrix.ipynb` (as another section) and remove  `Translation_Matrix_Revisit.ipynb`

menshikh-iv commented:
In general, LGTM for me, have you any suggestions @gojomo @piskvorky?

gojomo commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,329 @@
+
+"""
+Produce translation matrix to translate the word from one language to another language, using either
+standard nearest neighbour method or globally corrected neighbour retrieval method [1].
```
Could note that this is potentially more useful than just word-to-word translation; it'll work for any two sets of named-vectors where there are some paired-guideposts to learn the transformation.

gojomo commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,329 @@
+
+Train a model with e.g.::
+
+    >>> transmat.train(word_pair)
```
This API that supplies `word_pair` twice - both at construction & a later necessary `train()` step – seems odd to me, and in contrast to other gensim practice (where training can be combined with initialization, if training data is supplied.)

robotcator replied:
Done.

gojomo commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,329 @@
+
+    Attributes:
+        `mat` (ndarray): each row is the word vector of the lexicon
+        `index2word` (list): a list of lexicon
```
Terms like 'lexicon' unclear: is `index2word` a list of strings?

robotcator replied:
Yes, index2word is a list of words(strings).

gojomo commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,329 @@
+        return Space(mat, words)
+
+    def normalize(self):
+        """ normalized the word vector's matrix """
```
Capitalization, verb-tense: 'Normalize...'

gojomo commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,329 @@
+        Examples: [("one", "uno"), ("two", "due")]
+
+        Args:
+            `word_pair` (list): a list pair of words
```
As a list of many word-pairs, a more natural name for this parameter would be plural  `word_pairs`.

robotcator replied:
fixed

gojomo commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,329 @@
+        self.source_space = None
+        self.target_space = None
+
+    def train(self, word_pair):
```
This `word_pair` argument, which I noted seemed odd above, doesn't seem to be consulted at all in the following code. (Only the word-pairs supplied at initialization have any effect.)

robotcator replied:
As you said in the second comment, if training was combined with initialization, then this parameter would be used in `train` method.

gojomo replied:
Looking at the code body of the `train()` method, I see no references to this `word_pair` parameter. Hence, the parameter is unused and (unless the code changes) unnecessary here. If the intent was that this be an alternate way to supply the list-of-word-pairs, the code would have to change to reflect that intent.

robotcator replied:
Okay, I change the code where the `word_pairs` is optional  parameter in the initialization, and if provided, then `train` method will run followed.

gojomo commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,329 @@
+
+    def save(self, *args, **kwargs):
+        """
+        Save the model to file but ignoring the souce_space and target_space
```
`source_space`

robotcator replied:
Thanks, I will fix all the grammar errors and typos here and below.

gojomo commented on gensim/models/translation_matrix.py:
```diff
@@ -0,0 +1,329 @@
+
+    def apply_transmat(self, words_space):
+        """
+        mapping the source word vector to the target word vector using translation matrix
```
Capitalization, tense: "Map…"

gojomo commented on gensim/test/test_translation_matrix.py:
```diff
@@ -0,0 +1,125 @@
+        model = translation_matrix.TranslationMatrix(self.word_pair, self.source_word_vec, self.target_word_vec)
+        model.train(self.word_pair)
+
+        test_word_pair = [("one", "uno"), ("two", "due"), ("apple", "mela"), ("orange", "aranicione"), ("dog", "cane"), ("pig", "maiale"), ("cat", "gatto")]
```
Every one of these word-pairs is in the training mappings. A 'truer' test of 'translation' functionality would verify that using these known signpost-equivalences creates a mapping that's useful for getting *other* word-to-word translations approximately right.

gojomo commented on gensim/test/test_translation_matrix.py:
```diff
@@ -0,0 +1,125 @@
+        test_source_word, test_target_word = zip(*test_word_pair)
+        translated_words = model.translate(test_source_word, topn=3)
+
+        self.assertTrue("uno" in translated_words["one"])
```
This could be a loop driven by `test_word_pair`.

gojomo commented:
The constant stream of tiny tweak change-notifications, without any deeper review of the docs/naming issues I'd mentioned way-back, made me think this work was still far from review-readiness.  

I just now see it was it marked for `[MRG]` and merged in the last 10 hours. 

I think the API, naming, comments, and testing are all still a bit off. I've added some comments, but all of the code-doc-comments should get additional style/clarity proofreading.

robotcator commented:
@gojomo Thank you for reviewing, there still need more improvements. And I will fix the problem according to your comments.

robotcator commented:
@gojomo Hi, I open a new [PR](https://github.com/RaRe-Technologies/gensim/pull/1594) to fix the comment according to you. I change the `train` method to make it compulsory and in the `__init__` method to optional, but If the `word_pairs` is provided, the `train` method will also run. And I will make change in the notebook later.
